pylint/checkers/refactoring.py

```diff
@@ -311,6 +317,54 @@ def leave_functiondef(self, _):
         # new scope = reinitialize the stack of nested blocks
         self._nested_blocks = []
 
+    def visit_raise(self, node):
+        self._check_stop_iteration_inside_generator(node)
+
+    def _check_stop_iteration_inside_generator(self, node):
+        """
+        Check if an exception of type StopIteration is raised inside a generator
+        """
+        scope = node.scope()
+        if not isinstance(scope, astroid.FunctionDef) or not scope.is_generator():
+            return
+        if self._check_raise_inside_tryexcept_catching_stopiter(node):
+            return
+        exception_name = getattr(node.exc, 'name', '')
+        # Get all hand made exceptions defined inside the module
+        # that inherit from StopIteration
+        stopiteration_like_exc = [klass.name for klass in
+                                  node.root().nodes_of_class(astroid.ClassDef)
+                                  if self._check_exception_inherit_from_stopiteration(klass)]
+        if exception_name in ['StopIteration'] + stopiteration_like_exc:
+            self.add_message('stop-iteration-return', node=node)
+
+    @staticmethod
+    def _check_exception_inherit_from_stopiteration(exc):
+        """
+        Return True if the exception node in argument inherit from StopIteration
+        """
+        return any([_class.name == 'StopIteration' for _class in exc.mro()])
+
+    def _check_raise_inside_tryexcept_catching_stopiter(self, node):
+        """
+        Return True if the node is, or is nested inside, a TryExcept node
+        that handle StopIteration exception
+        """
+        if (isinstance(node, astroid.TryExcept)
+                and self._check_stop_iteration_catched(node)):
+            return True
+        elif getattr(node, 'parent', False):
+            return self._check_raise_inside_tryexcept_catching_stopiter(node.parent)
+        return False
+
+    @staticmethod
+    def _check_stop_iteration_catched(try_except):
+        """
```

Comment:
Ditto as well. If the docstring is too big, just break it in a summary, followed by an empty line and then by a more verbose documentation, if needed.